### PR TITLE
[AIRFLOW-2539] OS Environment variable for logging FILENAME_TEMPLATE

### PR DESCRIPTION
**Pull request by NielsZeilemaker**
Make sure you have checked _all_ steps below.

### JIRA
- [X] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-2359
    - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a JIRA issue.

### Description
- [X] Here are some details about my PR, including screenshots of any UI changes:
First check if an OS environment variable is defined before falling back to the default FILENAME_TEMPLATE

### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### Documentation
- [X] In case of new functionality, my PR adds documentation that describes how to use it.
    - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality
- [X] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by Fokko** [comment]
Thanks @NielsZeilemaker for this change. My suggestion would be to this through the config:
https://airflow.readthedocs.io/en/latest/howto/set-config.html

This will allow us to set it both in the config, and using environment variables.

Cheers

**Comment by NielsZeilemaker** [comment]
@Fokko I agree, however there is this comment:

https://github.com/NielsZeilemaker/incubator-airflow/blob/94407ff9bdee35e70bd385fdc752066d87e83e2c/airflow/config_templates/airflow_local_settings.py#L24

Which seems to want to split the logging config and the configuration, and hence I implemented it in this manner. I don't agree with the TODO btw.

**Comment by jgao54** [comment]
I am also leaning towards the idea of switching back to have logging config consolidated back into airflow config. From my observation, the separation seems to have caused quite a bit of confusion since it was implemented.

I recall the original intention of separating logging config and airflow config is to provide more tunable configs on the handler level.  However the handler configs actually are not quite flexible in airflow. For example, the way logging currently works in airflow **requires** root logger to set the handler to console/stdout so that [task handler can pick them up](https://github.com/apache/incubator-airflow/blob/master/airflow/task/task_runner/base_task_runner.py#L134-L140). Another reason is that any logging configurations does not allow them to be set via environment variables. Finally, webserver/scheduler logging config also follow a different pattern (use a separate config filepath that can be passed in the terminal when starting the webserver/scheduler). So right now technically we have 3 ways of configuring logging.

It would be more useful to move the common configs back to airflow.cfg. We could still potentially keep the airflow_local_settings.py to provide advanced logging config, but warn user that they should  'modify only if they know what they are doing' kinda thing.

**Comment by NielsZeilemaker** [comment]
My suggestion would be to modify the 'configuration.get' such that it accepts a default value, and hence we could use:

```
FILENAME_TEMPLATE = conf.get('core', 'LOGGING_FILENAME_TEMPLATE', default_value='{{ ti.dag_id }}/{{ ti.task_id }}/{{ ts }}/{{ try_number }}.log')
```

**Comment by jgao54** [comment]
I'm okay with this for now, just think at some later point we could make logging more intuitive. @Fokko ?

**Comment by NielsZeilemaker** [comment]
I talked to @fokko today, and he suggested moving tge default_value into the configuration file and keeping the configuration.get method as it is now. I'll update the pull request with these changes.

**Comment by Fokko** [comment]
Yes, logging is still an issue. Now again with the Kubernetes executor we need to rework it again. Would be nice to find a more robust way of logging.
Getting it from the config instead of the environment variable, allows us to have at least one single way of setting the variabele.

**Comment by NielsZeilemaker** [comment]
I've moved all global settings from the logging_config to the configuration file and added a section to the upgrading.md.

**Comment by NielsZeilemaker** [comment]
@Fokko any remarks on this PR?

**Comment by NielsZeilemaker** [comment]
@jgao54 any remarks on this PR?

**Review by Fokko — APPROVED** [review]

File: airflow/config_templates/default_airflow.cfg [hunk 1]
```diff
@@ -61,9 +62,15 @@ logging_level = INFO
...
 
 # Log format
+# we need to escape the curly braces by adding an additional curly brace
 log_format = [%%(asctime)s] {{%%(filename)s:%%(lineno)d}} %%(levelname)s - %%(message)s
 simple_log_format = %%(asctime)s %%(levelname)s - %%(message)s
 
+# Log filename format
+# we need to escape the curly braces by adding an additional curly brace
+log_filename_template = {{{{ ti.dag_id }}}}/{{{{ ti.task_id }}}}/{{{{ ts }}}}/{{{{ try_number }}}}.log
+log_processor_filename_template = {{{{ filename }}}}.log
```
Comment (danmactough) [inline airflow/config_templates/default_airflow.cfg:72]: @NielsZeilemaker I think these additional curly braces on LL71-72 (changing from double curly braces to 4 curly braces) are incorrect. In my environment, the break Jinja2 parsing. The additional curly braces on L540 (changing from single curly braces to double curly braces) are correct. I suspect this was just a mistake when making the changes, like maybe a wayward sed command. 😄

Comment (Fokko) [reply airflow/config_templates/default_airflow.cfg:72]: @danmactough Please try to add a test to capture this :)

Comment (danmactough) [reply airflow/config_templates/default_airflow.cfg:72]: @Fokko I think there already are tests that cover this:

- https://github.com/apache/incubator-airflow/blob/master/tests/utils/test_log_handlers.py#L181
- https://github.com/apache/incubator-airflow/blob/master/tests/utils/log/test_file_processor_handler.py#L35

The problem is that the config_templates don't match the tests.

Comment (danmactough) [reply airflow/config_templates/default_airflow.cfg:72]: @Fokko @NielsZeilemaker I finally figured out what is going on here, and it's not a bug. 😬 Those strings with additional curly braces end up getting parsed as template strings, formatted, and written out in `airflow.cfg`. So the source default config has to contain something like `{{{{ filename }}}}.log` in order to get the corresponding entry in `airflow.cfg` to be `{{ filename }}.log`. 

Sorry for the false alarm.

**Pull request closed without merging**